Improved decision on whether to download and/or overwrite media

lenaschimmel opened the pull request:
I've made a lot of changes to `download_file_if_larger`. I'm sorry for mixing multiple changes into a single PR, but they evolved in parallel. Here's what this changes:

 * Only download file if `byte_size_after != byte_size_before` because it seems *very* unlikely that two versions with same file size are actually different
 * Download file even if it is smaller *in bytes*, because less bytes may encode more / better pixels, as pointed out in #51
 * After downloading, check image size *in pixels* and use downloaded file if `pixels_after >= pixels_before`
 * Reordered contents of log line: index / count, filename, result (uppercase), explanation, downloaded bytes
 * Less code repetition for log outputs by prefixes and postfixes into `pref` and `post`
 * On error, print `{err}` (not sure if this is the best way to get a meaningful message out of the exception)
 * some more code cleanup

I'm using [imagesize_py](https://github.com/shibukawa/imagesize_py) which can determine the pixel size without loading the whole image, and supports many image formats. If called on a video file, it would *probably* return `-1, -1` which I try to handle in my code. But since I never encountered a video with `byte_size_after != byte_size_before`, this did not occur anyway. I'm going to test this later.

The output format has changed a lot and now looks like this (shortened). At `14/79` there is an example of a file with smaller byte size, but larger pixel size. At `37/79` there is an example of a video, but the pixel size was not inspected.

```
[...]
frequent. This script may not work if your account is protected. You may want to set it to public
before starting the download.

OK to start downloading? [y/n]y
  1/ 79 media/1357031742197104640-EtOs3ASWgAUCHa_.jpg: SKIPPED. Online version is same byte size, assuming same content. Not downloaded.
  2/ 79 media/1343985634613075968-EqbLdKdXIA02fXl.jpg: SUCCESS. New version is  50% larger in bytes and  88% larger in pixels. 0.2MB downloaded
  3/ 79 media/1343962182770421760-Eqa1eakW8AYm8dZ.jpg: SUCCESS. New version is  13% larger in bytes and  14% larger in pixels. 0.1MB downloaded
  4/ 79 media/1343962170422403075-EqazVseXEAAYI6R.png: SKIPPED. Online version is same byte size, assuming same content. Not downloaded.
[...]
 14/ 79 media/1237779856546574338-ES15j3iX0AwmlHo.png: SUCCESS. New version is -56% larger in bytes and  30% larger in pixels. 0.3MB downloaded
[...]
 37/ 79 media/864232876539351040-uTWbzDlB_rr2kkh0.mp4: SKIPPED. Online version is same byte size, assuming same content. Not downloaded./vid/534x360/uTWbzDlB_rr2kkh0.mp4...
[...]
 79/ 79 media/1249726622888005635-EVfrvEmXsAAwlVL.jpg: SUCCESS. New version is 105% larger in bytes and 125% larger in pixels. 0.2MB downloaded

79 of 79 tested media files are known to be the best-quality available.

Total downloaded: 14.4MB = 0.01GB
Time taken: 85s
Wrote log to media/download_log.txt
In case you set your account to public before initiating the download, do not forget to protect it again.
```

lenaschimmel commented:
If errors occur, it looks like that:

`3578/4692: media/1042087012990414848-rw2GV595Or-UUv8Z.mp4: FAIL. Media couldn't be retrieved from https://video.twimg.com/ext_tw_video/1042086982569086982/pu/vid/720x1280/rw2GV595Or-UUv8Z.mp4?tag=5 because of exception: 'content-length'`

And it seems to me like there should be a way to get more information about the problem than just `'content-length'`.

timhutton commented:
`New version is -56% larger in bytes and  30% larger in pixels`

Can we fix these negative cases to say maybe:

`New version is actually 56% smaller in bytes but 30% larger in pixels`

timhutton commented on parser.py:
```diff
@@ -194,32 +194,58 @@ def download_file_if_larger(url, filename, index, count, sleep_time):
     try:
         with requests.get(url, stream=True) as res:
             if not res.status_code == 200:
                 raise Exception('Download failed')
```
Now that we're printing the exceptions, please add the res.status code (and whatever else might be useful) here.

lenaschimmel replied:
I've done this now. I added status code, status text and text content of the response. Twitters responses for 404 are always empty, but I think it's nice to keep this code, so that content is displayed if Twitter will ever start to return error messages.

timhutton commented on parser.py:
```diff
@@ -194,32 +194,62 @@ def download_file_if_larger(url, filename, index, count, sleep_time):
+            if (byte_size_after != byte_size_before):
                 # Proceed with the full download
-                print(f'{index}/{count}: Downloading {url}...            ', end='\r')
+                print(f'{pref}Downloading {url}...', end='\r')
```
The extra spaces were because otherwise with `end=\r` on the previous print the shorter message wouldn't overwrite all of the previous message on this line.

Unrelated question: I wonder if the streaming is working as I thought. The 'Downloading' mesage is never shown for very long.

lenaschimmel replied:
Ah, I was wondering about this. I did not really understand how `print` and `logging.info` interacted (and still don't).

I also noticed a similar problem when a URL was very long, so that the line with `Downloading {url}...` was longer than the log line that came after it, so that a part of the URL was still visible like in line ` 37/ 79`:

```
 35/ 79 media/866363488830095361-DAXwkwOXgAEQCY4.jpg: SKIPPED. Online version is same byte size, assuming same content. Not downloaded.
 36/ 79 media/866363488830095361-DAXwmLZWAAA9jea.jpg: SKIPPED. Online version is same byte size, assuming same content. Not downloaded.
 37/ 79 media/864232876539351040-uTWbzDlB_rr2kkh0.mp4: SKIPPED. Online version is same byte size, assuming same content. Not downloaded./vid/534x360/uTWbzDlB_rr2kkh0.mp4...
 38/ 79 media/855595638846894080-C98CmmCW0AAIePq.jpg: SKIPPED. Online version is same byte size, assuming same content. Not downloaded.
 39/ 79 media/855046691480887296-C928ZraXcAA8c5b.jpg: SKIPPED. Online version is same byte size, assuming same content. Not downloaded.
```

So I can revert the missing space characters, but this will not fix the whole problem. Maybe there should be a separate issue for this?

> Unrelated question: I wonder if the streaming is working as I thought. The 'Downloading' message is never shown for very long.

I'm testing this on a 1 Gbit/s connection, so these small downloads are always faster than I can read :D

timhutton commented on parser.py:
```diff
@@ -194,32 +194,62 @@ def download_file_if_larger(url, filename, index, count, sleep_time):
                 # Proceed with the full download
-                print(f'{index}/{count}: Downloading {url}...            ', end='\r')
+                print(f'{pref}Downloading {url}...', end='\r')
                 with open(filename+'.tmp','wb') as f:
```
Suggest assigning filename+'.tmp' to a variable for reuse.

lenaschimmel commented:
I've addressed all your suggestions. I'm not sure if I may / should click on "Resolve conversation" or if you should be the one who decides if it's resolved.

timhutton commented:
@lenaschimmel There's signal either way, I'm not bothered. (My main emotion is being giddy with gratitude that people are working together to improve the code.) I don't know if there's a convention on github. (At work we have 'resolved' and 'closed' and typically the PR author will mark as resolved and then the comment author will mark as closed.)
